lib/src/dice_coefficient.dart

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:math' as math;
+
+/// Calculates the [Sørensen–Dice coefficient][1] for two strings.
+///
+/// This is done by converting the strings into bigrams (e.g. 'dart' becomes
+/// {'da', 'ar', 'rt}), and calculates the coefficient based on twice the
+/// number of common elements over the total number of elements.
+///
+/// In this implementation, identical bigrams that appear more than once in the
+/// string will be calculated accordingly (e.g. 'aaaa' is split into {'aa', 'aa'
+/// , 'aa'}), and each them a match is made, we strike off one of the bigrams.
+/// The effect of doing so increases the meaingfulness of the coefficient
+/// produced.
+///
+/// Our implementation also works with empty strings and one-character strigns.
+///
+/// **Example**
+/// ```dart
+/// diceCoefficient('', '')                                 /// 1.0
+/// diceCoefficient('', 'a')                                /// 0.0
+/// diceCoefficient('a', 'a')                               /// 1.0
+/// diceCoefficient('a', 'b')                               /// 0.0
+/// diceCoefficient('aa', 'aaa')                            /// 0.6666667
+/// diceCoefficient('aaaa', 'aaaaa')                        /// 0.857142
+/// diceCoefficient('night', 'nacht')                       /// 0.25
+/// diceCoefficient('dev_dependencies', 'dev-dependencies') /// 0.8666666
+/// diceCoefficient('dependency', 'dependencies')           /// 0.8
+/// ```
+///
+/// [1]: https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient
+double diceCoefficient(String string1, String string2) {
```

Comment:
Perhaps there is an argument to be made that we should investigate if this is the best to detect likely typos.
At-least we should look at what other fuzzy matching applications exists out there.

Reply:
Just tested with a levenshtein distance implementation, and the results are at: https://gist.github.com/walnutdust/5d0f151025fc986c5606d87fa15d5eb5.

From this it does seem like levenshtein would be the better way to go, and I can code it up once we can confirm this!

Reply:
I agree this is much better let's go with experiment (2) or (3) from:
https://gist.github.com/walnutdust/5d0f151025fc986c5606d87fa15d5eb5